### ISSUE TITLE
USB serial (and other exposed devices) not available to the deck user.

### ISSUE THREAD

**Issue by oemsysadm** [ValveSoftware/SteamOS#1635]
#### Your system information

* Steam client version: Latest Stable
* SteamOS version: Latest Stable
* Opted into Steam client beta?: [Yes/No] NO
* Opted into SteamOS beta?: [Yes/No] No
* Have you checked for updates in Settings > System?: [Yes/No] N/A

#### Please describe your issue in as much detail as possible:
Describe what you _expected_ should happen and what _did_ happen.

I am trying to run software, currently installed as a flatpak (locally built) which need access to a USB serial port to interact with a connected device (Dirtywave M8 Headless.) Running the flatpak as root (sudo -i; flatpak run ...) works, but not as the deck user because that user does not have access to the serial port exposed by the associated device.

The associated device also has audio/midi over USB, and these are also not working correctly. Audio plays but there is no way to plumb the audio to the speakers that I can find.

I would argue this is at least gaming-adjacent and with more and more applications appearing on Steam, some audio applications might be locked out for not having access to USB hardware, such as midi adapters, audio interfaces, etc.

On other Arch Linux derivative distros, the user in question simply has to be a member of various groups, such as the "uucp" group.

#### Steps for reproducing this issue:

I would assume run any terminal program and use a USB based serial adapter, audio interface, or similar device.

**Comment by kisak-valve** [comment]
Hello @oemsysadm, this reads more like a Flatpak issue than a SteamOS issue. Maybe give https://stackoverflow.com/questions/41252334/accessing-serial-ports-with-an-application-runs-in-flatpak a try?

**Comment by oemsysadm** [comment]
Alas same issue with the bare bimary. The user needs to be in the uucp group.

**Comment by oemsysadm** [comment]
Yes, or in /etc/group. But these things get reset when the OS is updated on the steam deck. I'm looking to have better support for external usb hardware, bluetooth, midi, etc baked into the distro so we can distribute applications in steam or via flathub that "just work" for these purposes, hence this github issue.

**Comment by bertogg** [comment]
> But these things get reset when the OS is updated on the steam deck.

You can install custom udev rules in `/etc/udev/rules.d/` and keep them during OS updates, see https://github.com/ValveSoftware/SteamOS/issues/1601#issuecomment-2346185382

> I'm looking to have better support for external usb hardware, bluetooth, midi, etc baked into the distro 

What is the USB id of the hardware (`lsusb`) and what is the name of the device files under `/dev` that you need access to?

**Comment by oemsysadm** [comment]
Udev is fine the device (essentially a teensy) shows and works. The issue is the deck users access to various hardware, like serial ports, midi, etc. Typically governed by group affiliation in Arch. Being able to permanently or temporarily (until an OS update) edit OS files doesn't address the need for these permissions in a painless way when using, for instance, music editors purchased through Steam or installed via Glathub, that need access to outboard hardware. I assume there may be similar issues when using specialized hardware connected for graphics applications, etc.

**Comment by bertogg** [comment]
Adding a user to a group statically in `/etc/group` in order to have access to a USB device is not a recommended practice.

Instead, udev is used to give access dynamically to users who have logged in¹. If this does not work automatically then it means that there is a udev rule missing for that device.

It is common that libraries or apps ship udev rules for the devices that they handle if there isn't one already, see for example https://github.com/Nitrokey/libnitrokey/blob/v3.8/data/41-nitrokey.rules

---------
¹ More information here:

https://enotty.pipebreaker.pl/2012/05/23/linux-automatic-user-acl-management/

https://wiki.archlinux.org/title/Udev
